kbase-extension/static/kbase/js/widgets/function_output/kbaseFeatureSet.js

```diff
@@ -62,162 +64,270 @@ define (
-                })
+            return this.genomeSearchAPI.callFunc('search', [{
+                ref: genome_ref,
+                structured_query: query,
+                sort_by: [['contig_id',1]],
+                start: 0,
+                limit: limit
+            }])
+                .then(([result]) => {
+                    return result;
+                });
         },
 
-
         genomeLookupTable: null, // genomeId: { featureId: indexInFeatureList }
         genomeObjectInfo: null, //{},
         featureTableData: null, // list for datatables
 
         getGenomeData: function() {
-            var self = this;
-            self.genomeLookupTable = {};
-            self.genomeObjectInfo = {};
-            self.featureTableData = [];
-            for(var gid in self.features) {
-                var query = {"feature_id": self.features[gid]}
-                self.search(gid, {"feature_id": self.features[gid]}, self.features[gid].length)
-                    .then(function(results) {
-                        for (var f in results.features) {
-                            var feature = results.features[f];
-                            self.featureTableData.push(
+            if (Object.keys(this.features).length === 0) {
+                this.loading(false);
+                this.showMessage('This feature set is empty.');
+                return Promise.resolve();
+            }
+
+            this.genomeLookupTable = {};
+            this.genomeObjectInfo = {};
+            this.featureTableData = [];
+            // We fetch the data with nested promises, to preserve order and parallelize the requests.
+            // Then the results are dissected and used to update structures in this object, and finally
+            // invoking a render.
+            //
+            // "this.features" is an object whose keys are genome object references (gid), and
+            // whose values are lists of feature ids (fid).
+            // For each pair of gid and fid we need to fetch the genome object information from the workspace
+            // and the feature info from the genome search api.
+            // To keep this all sorted out, and in the original order, we do this as an array of promises for
+            // each pair, and each pair itself is an array of the actual api calls which are themselves promises.
+
+            // get the object info for all feature-containing objects.
+            const objectRefs = Object.keys(this.features);
+            return this.workspace.callFunc('get_object_info3', [{
+                objects: objectRefs.map((ref) => {
+                    return {
+                        ref
+                    };
+                })
+            }])
+                .then(([result]) => {
+                    return Promise.all(result.infos.map((info) => {
+                      
+                        const [/* typeModule */, typeName, /*typeVersionMajor*/, /*typeVersionMinor*/] = info[2].split(/[.-]/);
```

Comment:
Can this be cleaned up with comments removed?

Reply:
I put those in to document the empty slots, can be removed of course, just leaving empty array slots.